# 📊 Exploratory Data Analysis (EDA)

## School Bullying Analysis using the CRDC 2021–22 Dataset

**Notebook:** 03 – Exploratory Data Analysis

**Author:** Yoldas Erdem

**Project:** Data Analytics & AI Bootcamp Capstone

---

# 📖 Introduction

Following the completion of the ETL pipeline, this notebook marks the beginning of the analytical phase of the project.

The cleaned CRDC Harassment and Bullying dataset is explored to better understand its structure, assess its overall quality, and identify meaningful patterns before conducting SQL analysis and building interactive Tableau dashboards.

Rather than testing predefined hypotheses, the purpose of this notebook is to become familiar with the data, uncover potential relationships, and identify questions worth investigating in greater depth.

Exploratory Data Analysis (EDA) serves as a critical bridge between data preparation and statistical analysis. The findings documented throughout this notebook will directly inform the SQL queries, dashboard design, and final presentation.

---

# 🎯 Objectives

This notebook aims to:

- Understand the overall structure of the cleaned dataset.
- Verify that the ETL process produced a reliable analytical dataset.
- Explore the distribution of bullying allegations across different categories.
- Investigate demographic patterns among affected students.
- Examine disciplinary actions reported by schools.
- Identify geographic trends across states and school districts.
- Generate analytical questions for the SQL phase.
- Define potential visualizations for the Tableau dashboard.

---

# 🗺️ Notebook Roadmap

The exploratory analysis is organized into three major sections.

## Section 1 — Dataset Understanding

Gain an understanding of the dataset before beginning any analysis.

Topics include:

- Dataset dimensions
- Data types
- Missing values
- Duplicate records
- Metadata overview
- Summary statistics

---

## Section 2 — Data Exploration

Explore the main components of the CRDC dataset and identify meaningful patterns.

Topics include:

- School metadata
- Bullying allegations
- Reported students
- Disciplinary actions
- Geographic comparisons
- Distribution analysis
- Correlation analysis

---

## Section 3 — Key Findings

Summarize the most important observations and prepare for the analytical phase of the project.

Topics include:

- Key insights
- Potential limitations
- Questions for SQL analysis
- Ideas for Tableau dashboards

---

# 🔄 Analytics Workflow

This notebook represents the third stage of the overall project workflow.

```

Raw CRDC Dataset
        │
        ▼
ETL Pipeline
        │
        ▼
Clean Analytical Dataset
        │
        ▼
📍 Exploratory Data Analysis
        │
        ▼
PostgreSQL Database
        │
        ▼
SQL Analysis
        │
        ▼
Tableau Dashboard
        │
        ▼
Executive Presentation

```

---

# 🚀 Let's Begin

The analysis starts with a high-level assessment of the cleaned dataset to ensure its structure, completeness, and quality before exploring individual variables in greater detail.

# 1. Dataset Understanding

## Purpose

Before exploring individual variables, it is important to understand the overall structure and quality of the dataset.

This section provides a high-level overview of the cleaned CRDC dataset by examining its dimensions, data types, completeness, and basic descriptive statistics. These initial checks confirm that the ETL process produced a reliable analytical dataset and establish a foundation for the exploratory analyses that follow.

---

### Analysis Checklist

- Load the cleaned dataset
- Review dataset dimensions
- Inspect data types
- Examine missing values
- Verify duplicate records
- Review summary statistics
- Inspect the first observations

# 1.1 Load the Clean Dataset

## Purpose

The exploratory analysis begins by loading the cleaned dataset produced during the ETL pipeline.

Unlike the original CRDC dataset, this version contains standardized variable names, preserved identifier fields, and CRDC reserve codes converted to missing values (`NULL`). It serves as the primary analytical dataset used throughout the remainder of this project.

After loading the dataset, a preview of the first observations is displayed to verify that the data was imported successfully.

In [1]:
# =====================================================
# Import Required Libraries
# =====================================================

from pathlib import Path
import pandas as pd

In [2]:
# =====================================================
# Project Configuration
# =====================================================

# Define the project root directory
project_root = Path.cwd().parent

# Path to the cleaned dataset created during the ETL pipeline
clean_data = project_root / "data" / "processed" / "crdc_bullying_clean.csv"

In [3]:
# =====================================================
# Load the Clean Dataset
# =====================================================

df = pd.read_csv(
    clean_data,
    dtype={
        "district_id": str,
        "school_key": str
    },
    low_memory=False
)

# Display the first five records
df.head()

,state_code,state,district_id,district,school_id,school,school_key,juvenile_justice_school,allegation_sex,allegation_orientation,...,disciplined_disability_multi_race_male,disciplined_disability_multi_race_female,total_disciplined_disability_male,total_disciplined_disability_female,disciplined_disability_english_learner_male,disciplined_disability_english_learner_female,disciplined_disability_idea_male,disciplined_disability_idea_female,disciplined_disability_section_504_male,disciplined_disability_section_504_female
0,AL,ALABAMA,0100002,Alabama Youth Services,99995,AUTAUGA CAMPUS,010000299995,Yes,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,AL,ALABAMA,0100005,Albertville City,870,Albertville Middle School,010000500870,No,2.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,AL,ALABAMA,0100005,Albertville City,871,Albertville High School,010000500871,No,4.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,AL,ALABAMA,0100005,Albertville City,879,Albertville Intermediate School,010000500879,No,1.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,AL,ALABAMA,0100005,Albertville City,889,Albertville Elementary School,010000500889,No,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


# 1.2 Dataset Overview

## Purpose

Before examining individual variables, it is essential to understand the overall structure of the dataset.

This section provides a high-level overview, including the number of observations, variables, data types, and memory usage. These statistics help verify that the dataset has been loaded correctly and provide context for the analyses that follow.

In [4]:
# =====================================================
# Dataset Dimensions
# =====================================================

print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]}")

Rows: 98,010
Columns: 159


In [5]:
# =====================================================
# Dataset Information
# =====================================================

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 98010 entries, 0 to 98009
Columns: 159 entries, state_code to disciplined_disability_section_504_female
dtypes: float64(151), int64(1), str(7)
memory usage: 118.9 MB


In [6]:
# =====================================================
# Summary Statistics
# =====================================================

df.describe()

,school_id,allegation_sex,allegation_orientation,allegation_race,allegation_disability,allegation_religion,allegation_atag,allegation_budd,allegation_cath,allegation_east,...,disciplined_disability_multi_race_male,disciplined_disability_multi_race_female,total_disciplined_disability_male,total_disciplined_disability_female,disciplined_disability_english_learner_male,disciplined_disability_english_learner_female,disciplined_disability_idea_male,disciplined_disability_idea_female,disciplined_disability_section_504_male,disciplined_disability_section_504_female
count,98010.000000,91902.000000,91902.000000,91902.000000,91902.000000,91902.000000,91777.000000,91777.000000,91777.000000,91777.000000,...,95738.000000,95738.000000,95738.000000,95738.000000,95759.000000,95760.000000,95741.000000,95741.000000,95741.000000,95741.000000
mean,3926.152944,0.641738,0.251453,0.456922,0.172564,0.029346,0.000338,0.000022,0.000131,0.000044,...,0.007823,0.003677,0.138012,0.051505,0.012030,0.004094,0.035982,0.009296,0.005828,0.001734
std,9765.355942,5.135084,4.564510,4.917722,4.562140,0.367293,0.027017,0.004668,0.013203,0.006602,...,0.242827,0.178743,2.668595,1.803655,0.369702,0.209039,0.581996,0.274646,0.187302,0.079802
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,753.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,1785.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
75%,4184.750000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
max,99999.000000,1133.000000,1133.000000,1133.000000,1133.000000,43.000000,5.000000,1.000000,2.000000,1.000000,...,50.000000,37.000000,573.000000,390.000000,57.000000,36.000000,86.000000,47.000000,46.000000,14.000000


# 1.3 Variable Overview

## Purpose

Understanding the variables available in the dataset is an important first step in exploratory analysis.

This section categorizes the variables according to their analytical purpose. Organizing the dataset into logical groups provides a clearer understanding of its structure and establishes the foundation for the analyses performed throughout the remainder of this notebook.

The CRDC Harassment and Bullying dataset can be divided into four main categories:

- **Metadata** – School and district identifiers.
- **Allegations** – Reported harassment and bullying allegations.
- **Reported Students** – Students reported as affected by bullying.
- **Disciplined Students** – Students receiving disciplinary actions following reported incidents.

In [8]:
# =====================================================
# Categorize Variables
# =====================================================

# Metadata variables
metadata_cols = [
    "state_code",
    "state",
    "district_id",
    "district",
    "school_id",
    "school",
    "school_key",
    "juvenile_justice_school"
]

# Allegation variables
allegation_cols = [
    col for col in df.columns
    if col.startswith("allegation_")
]

# Reported student variables
reported_cols = [
    col for col in df.columns
    if col.startswith("reported_") or col.startswith("total_reported_")
]

# Disciplined student variables
disciplined_cols = [
    col for col in df.columns
    if col.startswith("disciplined_") or col.startswith("total_disciplined_")
]

In [9]:
# =====================================================
# Variable Category Summary
# =====================================================

variable_summary = pd.DataFrame({
    "Variable Category": [
        "Metadata",
        "Bullying Allegations",
        "Reported Students",
        "Disciplined Students"
    ],
    "Number of Variables": [
        len(metadata_cols),
        len(allegation_cols),
        len(reported_cols),
        len(disciplined_cols)
    ]
})

variable_summary.loc[len(variable_summary)] = [
    "Total",
    variable_summary["Number of Variables"].sum()
]

variable_summary

,Variable Category,Number of Variables
0,Metadata,8
1,Bullying Allegations,19
2,Reported Students,66
3,Disciplined Students,66
4,Total,159


### Interpretation

The cleaned dataset contains **159 analytical variables**, organized into four logical categories.

The majority of variables describe either **students reported as affected by bullying (66 variables)** or **students receiving disciplinary actions (66 variables)**. These two groups account for more than 80% of all variables and represent the primary focus of the analytical phase.

A smaller group of **19 allegation variables** records the different types of harassment and bullying reported by schools, while **8 metadata variables** provide the geographic and institutional context required for comparisons across schools, districts, and states.

This categorization provides a clear roadmap for the exploratory analysis that follows.